### ISSUE TITLE
Bug: [@typescript-eslint/require-await] False positive when implementing async interface with sync function

### ISSUE THREAD

**Issue by panstromek** [typescript-eslint/typescript-eslint#7450]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=LAKAlgdgLgpgTgMwIYGMYAICCBnAnhFAZQAsB7Ad3QG9R071szyAudACgEp0BeAPnQAKcUgFsw2GAB4AbqTAATXqAC%2BoUAgCuBKGFIR0I0igDWOfCgBiETqzMESFarXpwYUDXH00Q9X%2BiR4BAxMnE4%2BfhEoetikADYwAHSxpADmbADkEKQADukczhG%2Bru6eBX6q4XQVFUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRAR1ko9oEMA7v0r50URNGgB7aJHBgAviEVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
interface AsyncShow {
    show: () => Promise<void>
}

function mockAsyncFn(): AsyncShow {
    return {
        async show() {
            console.log('nop')
            return
        }
    }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/require-await": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error.

### Actual Result

Error on `show` implementation: Async method 'show' has no 'await' expression.

### Additional Info

Since the method just implement an interface (which might be out of scope), it shouldn't require await keyword if it is not necessary for this specific implementation.

The way to fix this is to use `Promise.resolve()` inside non-async function, but that's unnecessarily verbose.

**Comment by bradzacher** [comment]
See discussion on #7536
After consideration - this is working as expected.

Relying on `async` to lazily satisfy a promise-based contract is generally considered a code smell and is one of the reasons this rule exists.
Ignoring this case would leave the code ambiguous - "did the author forget an `await`? Did they msirwite the code? Did they intentionally use `async` to wrap the return in a promise?"

Instead the rule is encouraging you to be explicit and clearly mark your return with a promise (eg `Promise.resolve()`) which unambiguously signals that the imementatikn intentionally does not contain asynchronous code.
